File: Errors/errors/ErrorCalculator.go
package errors

import "math"

func AbsoluteError(original float64, estimated float64) float64 {
	return math.Abs(original-estimated)
}

func RelativeError(original float64, estimated float64) float64 {
	return AbsoluteError(original, estimated) / math.Abs(original)
}

/*
	result < 1 ---> f reduce the error
	result = 1 ---> f preserve the error
	result > 1 ---> f increase the error
*/
func FunctionEffectOnResult(original []float64, estimated []float64, f func([]float64)(float64)) float64{
	var inputRelativeError float64 = sumRelativeError(original, estimated)

	var outputRelativeError float64 = RelativeError(f(original), f(estimated))

	return outputRelativeError / inputRelativeError
}

func sumRelativeError(original []float64, estimated []float64) float64 {
	var sum float64 = 0

	for index := range original {
		sum += RelativeError(original[index], estimated[index])
	}

	return sum
}

File: Errors/ErrorCalculator.py
from typing import Callable, List
import math


def absoluteError(original: float, estimated: float) -> float:
    return abs(original - estimated)


def relativeError(original: float, estimated: float) -> float:
    return absoluteError(original, estimated) / abs(original)


# result < 1 ---> f reduce the error
# result = 1 ---> f preserve the error
# result > 1 ---> f increase the error
def functionEffectOnResult(original: List[float], estimated: List[float],
                           f: Callable[[List[float]], float]) -> float:
    
    if len(original) != len(estimated):
        raise Exception("the inputs are not in the same size")
    
    inputRelativeError: float = math.fsum([relativeError(org, est) for (org, est) in zip(original, estimated)])

    outputRelativeError: float = relativeError(f(original), f(estimated))

    return outputRelativeError / inputRelativeError



